### PR TITLE
kubectl: implement kubectl apply --prune

### PR DESCRIPTION
**Pull request by mikedanese**
Closes https://github.com/kubernetes/kubernetes/issues/19805

Depends on #32599 

@errordeveloper @lukemarsden

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33075)

### CODE REVIEW

**Comment by errordeveloper** [comment]
Awesome! Thanks, @mikedanese.

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -71,15 +84,19 @@ func NewCmdApply(f *cmdutil.Factory, out io.Writer) *cobra.Command {
...
 	}
 
 	usage := "that contains the configuration to apply"
-	cmdutil.AddFilenameOptionFlags(cmd, options, usage)
+	cmdutil.AddFilenameOptionFlags(cmd, &options.FilenameOptions, usage)
 	cmd.MarkFlagRequired("filename")
 	cmd.Flags().Bool("overwrite", true, "Automatically resolve conflicts between the modified and live configuration by using values from the modified configuration")
+	cmd.Flags().BoolVar(&options.Prune, "prune", false, "Automatically delete missing objects from supplied config")
+	cmd.Flags().BoolVar(&options.Cascade, "cascade", true, "Only relevant during a prune. If true, cascade the deletion of the resources managed by pruned resources (e.g. Pods created by a ReplicationController).")
+	cmd.Flags().IntVar(&options.GracePeriod, "grace-period", -1, "Period of time in seconds given to pruned resources to terminate gracefully. Ignored if negative.")
```
Comment (errordeveloper) [inline pkg/kubectl/cmd/apply.go]: Would it be more user-friendly to say "Wait forever if negative."? (Given I even understood this correctly).

File: pkg/kubectl/cmd/apply.go [hunk 2]
```diff
@@ -71,15 +84,19 @@ func NewCmdApply(f *cmdutil.Factory, out io.Writer) *cobra.Command {
...
 		},
 	}
 
 	usage := "that contains the configuration to apply"
-	cmdutil.AddFilenameOptionFlags(cmd, options, usage)
+	cmdutil.AddFilenameOptionFlags(cmd, &options.FilenameOptions, usage)
 	cmd.MarkFlagRequired("filename")
 	cmd.Flags().Bool("overwrite", true, "Automatically resolve conflicts between the modified and live configuration by using values from the modified configuration")
+	cmd.Flags().BoolVar(&options.Prune, "prune", false, "Automatically delete missing objects from supplied config")
+	cmd.Flags().BoolVar(&options.Cascade, "cascade", true, "Only relevant during a prune. If true, cascade the deletion of the resources managed by pruned resources (e.g. Pods created by a ReplicationController).")
```
Comment (errordeveloper) [inline pkg/kubectl/cmd/apply.go]: When a user may need to use this? May be `--no-cascaded-prune` would be a better name for this.

File: pkg/kubectl/resource/builder.go [hunk 3]
```diff
@@ -538,6 +538,41 @@ func (b *Builder) visitorResult() *Result {
 		b.selector = labels.Everything()
 	}
 
+	// visit items specified by paths
```
Comment (errordeveloper) [inline pkg/kubectl/resource/builder.go]: Is this code simply moved from down below, or it also contains some changes? May be good idea to make a separate PR for moving this first. Also, this looks like a ginormous func and some parts of it can be broken out,

File: pkg/kubectl/cmd/apply.go [hunk 4]
```diff
@@ -191,11 +233,112 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
...
-
 	if count == 0 {
 		return fmt.Errorf("no objects passed to apply")
 	}
 
+	if !options.Prune {
+		return nil
+	}
+
+	selector, err := labels.Parse(options.Selector)
```
Comment (errordeveloper) [inline pkg/kubectl/cmd/apply.go]: Something like `return visitAndPruneIfNeeded(options.Selector)` could be easier to read, wdyt?

File: pkg/kubectl/cmd/apply.go [hunk 5]
```diff
@@ -191,11 +233,112 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
...
+	for _, m := range visitedNonNamespacedRESTMappings {
+		if err := p.prune(api.NamespaceNone, m, shortOutput); err != nil {
+			return fmt.Errorf("error pruning objects: %v", err)
+		}
+	}
+
+	return nil
+}
+
+type pruner struct {
```
Comment (errordeveloper) [inline pkg/kubectl/cmd/apply.go]: To me it'd be perfectly reasonable to put have this in a separate file, when I was trying to read `apply.go` last time, it seemed rather big and not quite structured.

**Review by errordeveloper — COMMENTED** [review]
Few nits on UX and readability...

File: pkg/kubectl/resource/builder.go (hunk 3 above)
Comment (mikedanese) [inline pkg/kubectl/resource/builder.go]: https://github.com/kubernetes/kubernetes/pull/32599

File: pkg/kubectl/cmd/apply.go (hunk 1 above)
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: It's actually defaulted by the apiserver, not wait forever but I can say something to that effect

File: pkg/kubectl/cmd/apply.go (hunk 2 above)
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: --cascade is already used in multiple other commands in kubectl. I think continuity of experience outweighs a nice descriptive flag. 

> When a user may need to use this?

This is the default behavior. When we prime a deployment, we should also prune it's replicasets or they will become orphaned and unmanaged.

**Comment by mikedanese** [comment]
@kubernetes/kubectl

**Comment by kargakis** [comment]
cc: @kubernetes/kubectl

File: pkg/kubectl/cmd/apply.go (hunk 4 above)
Comment (soltysh) [inline pkg/kubectl/cmd/apply.go]: :+1: for separate function, but I'd rather have it return err and check it here, than what you propose.

**Comment by deads2k** [comment]
@fabianofranz fyi

File: pkg/kubectl/cmd/apply.go [hunk 6]
```diff
@@ -183,6 +219,12 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
 			}
 		}
 
+		if uid, err := info.Mapping.UID(info.Object); err != nil {
+			glog.Infof("here")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: remove.

File: pkg/kubectl/cmd/apply.go [hunk 7]
```diff
@@ -191,11 +233,112 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
...
+	}
+	objs, err := meta.ExtractList(objList)
+	if err != nil {
+		return err
+	}
+
+	for _, obj := range objs {
+		uid, err := mapping.UID(obj)
+		if err != nil {
+			glog.Infof("here")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: +1

File: pkg/kubectl/cmd/apply.go [hunk 8]
```diff
@@ -191,11 +233,112 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
...
+			return err
+		}
+		if p.visitedUids.Has(string(uid)) {
+			continue
+		}
+
+		name, err := mapping.Name(obj)
+		if err != nil {
+			glog.Infof("here")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/apply.go]: +1

Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: fixed

File: pkg/kubectl/cmd/apply.go (hunk 7 above)
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: fixed

File: pkg/kubectl/cmd/apply.go (hunk 6 above)
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: fixed

File: pkg/kubectl/cmd/apply.go [hunk 9]
```diff
@@ -158,6 +188,12 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
 			if err := createAndRefresh(info); err != nil {
 				return cmdutil.AddSourceToErr("creating", info.Source, err)
 			}
+			if uid, err := info.Mapping.UID(info.Object); err != nil {
+				glog.Infof("here")
```
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: fixed

**Comment by MrHohn** [comment]
It seems like the `kubectl apply --prune` with current commits will mistakenly prune the pods created by deployment/RC? As below:
```
$ kubectl apply -f apply-test.yaml
pod "jessie2" created
deployment "my-nginx1" created
replicationcontroller "my-nginx2" created
$ kubectl get pods
NAME                         READY     STATUS    RESTARTS   AGE
jessie1                      1/1       Running   0          10s
jessie2                      1/1       Running   0          10s
my-nginx1-3491544093-oz0sc   1/1       Running   0          10s
my-nginx1-3491544093-yjbix   1/1       Running   0          10s
my-nginx2-dhi03              1/1       Running   0          10s
my-nginx2-r51j1              1/1       Running   0          10s
# change nothing
$ kubectl apply -f apply-test.yaml --prune
pod "jessie1" configured
pod "jessie2" configured
deployment "my-nginx1" configured
replicationcontroller "my-nginx2" configured
pod "my-nginx1-3491544093-oz0sc" pruned
pod "my-nginx1-3491544093-yjbix" pruned
pod "my-nginx2-dhi03" pruned
pod "my-nginx2-r51j1" pruned
$ kubectl get pods
NAME                         READY     STATUS    RESTARTS   AGE
jessie1                      1/1       Running   0          48s
jessie2                      1/1       Running   0          48s
my-nginx1-3491544093-ysfta   1/1       Running   0          7s
my-nginx1-3491544093-zm6ue   1/1       Running   0          7s
my-nginx2-41f0p              1/1       Running   0          7s
my-nginx2-s5n06              1/1       Running   0          7s
```

Or should the proper way to do it be to utilize the labels?

**Comment by mikedanese** [comment]
@MrHohn can you show me the config you used? You likely want a label selector that only matches the top level objects you want to manage. e.g.

```
$ kubectl apply -l k8s-managed-addon=true -f apply-test.yaml
```
And add k8s-managed-addon=true to the labels of the top level objects you want to manage *but not* to the pods.

**Comment by MrHohn** [comment]
Sure. Config as below:
```
apiVersion: v1
kind: Pod
metadata:
  name: jessie1
spec:
  containers:
  - name: jessie
    image: girishkalele/jessie-full:1.0
    command: ["sh", "-c"]
    args: ["while true ; do sleep 3600 ; done"]
  restartPolicy: Never
---
apiVersion: v1
kind: Pod
metadata:
  name: jessie2
spec:
  containers:
  - name: jessie
    image: girishkalele/jessie-full:1.0
    command: ["sh", "-c"]
    args: ["while true ; do sleep 3600 ; done"]
  restartPolicy: Never
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: my-nginx1
spec:
  replicas: 2
  template:
    metadata:
      labels:
        run: my-nginx1
    spec:
      containers:
      - name: my-nginx
        image: nginx
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: ReplicationController
metadata:
  name: my-nginx2
spec:
  replicas: 2
  template:
    metadata:
      labels:
        run: my-nginx2
    spec:
      containers:
      - name: my-nginx
        image: nginx
        ports:
        - containerPort: 80

```

**Comment by MrHohn** [comment]
I see now. But why not making the label mandatory for --prune? As if label is not set it will prune innocence resources?

**Comment by mikedanese** [comment]
Empty label selector is valid (matches all) which is the default. I think we should add a flag `--all` to this before merging so users don't shot themselves in the foot.

File: pkg/kubectl/cmd/apply.go [hunk 10]
```diff
@@ -191,11 +230,110 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *re
...
+			if err := p.prune(n, m, shortOutput); err != nil {
+				return fmt.Errorf("error pruning objects: %v", err)
+			}
+		}
+	}
+	for _, m := range visitedNonNamespacedRESTMappings {
+		if err := p.prune(api.NamespaceNone, m, shortOutput); err != nil {
+			return fmt.Errorf("error pruning objects: %v", err)
+		}
+	}
```
Comment (MrHohn) [inline pkg/kubectl/cmd/apply.go]: Is using `visitedRESTMappings` here enough to cover all cases?

Given a YAML file which has a ReplicationController before, and now it is replaced with a Deployment. If we need to prune the RC,  I think this logic will fail? Because resource kind RC will not be visited this time -> RCs will not be listed by pruner -> fail to prune existing RC?

Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: Good point. I think we might need to add a --types flag to explicitly override visitedRESTMappings.

**Comment by pwittrock** [comment]
General comment - this will result in the canonical result of apply to no longer live in source control since the label selectors define what will be deleted.  Seems like this is very error prone - running apply on a subdirectory with prune will cause it to delete all sorts of things.  At a minimum I think we need to print out what will be deleted and force the user to confirm.

---

Review status: 0 of 3 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/cmd/apply.go, line 94 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL-CmSt-aBjoM1WYZi:-KTL-CmSt-aBjoM1WYZj:bgqcmdq) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L94)):*
> ```Go
> 	cmd.MarkFlagRequired("filename")
> 	cmd.Flags().Bool("overwrite", true, "Automatically resolve conflicts between the modified and live configuration by using values from the modified configuration")
> 	cmd.Flags().BoolVar(&options.Prune, "prune", false, "Automatically delete missing objects from supplied config")
> ```

Help message is a bit confusing - I read it as it modifies the configs to remove objects that are not found in the sever.  Maybe "Automatically delete resource objects that do not appear in the configs."

---

*[pkg/kubectl/cmd/apply.go, line 95 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL-dV0KuBMMOvUyhxR:-KTL-dV0KuBMMOvUyhxS:b-7yw4dh) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L95)):*
> ```Go
> 	cmd.Flags().Bool("overwrite", true, "Automatically resolve conflicts between the modified and live configuration by using values from the modified configuration")
> 	cmd.Flags().BoolVar(&options.Prune, "prune", false, "Automatically delete missing objects from supplied config")
> 	cmd.Flags().BoolVar(&options.Cascade, "cascade", true, "Only relevant during a prune. If true, cascade the deletion of the resources managed by pruned resources (e.g. Pods created by a ReplicationController).")
> ```

Ug.  Does the garbage collector handle this for us yet?

---

*[pkg/kubectl/cmd/apply.go, line 98 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL-pE24pjujI6hatcx:-KTL-pE24pjujI6hatcy:b-1lzue8) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L98)):*
> ```Go
> 	cmd.Flags().IntVar(&options.GracePeriod, "grace-period", -1, "Period of time in seconds given to pruned resources to terminate gracefully. Ignored if negative.")
> 	cmdutil.AddValidateFlags(cmd)
> 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", "", "Selector (label query) to filter on")
> ```

What does an empty value mean?  Delete everything by default?  We probably shouldn't permit using the default in that case

---

*[pkg/kubectl/cmd/apply.go, line 318 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL3J1A4D8jRFTTxKNy:-KTL3J1BVUOF0lLWpxeb:b-tzen04) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L318)):*
> ```Go
> 			return err
> 		}
> 		if p.cascade {
> ```

Consider pulling this whole logic into its own function - e.g. 'Delete' and it figures out to delete the thing

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-:-KTL3YLmANxot9OkXzMe:bb3o0kp)*

**Comment by pwittrock** [comment]
Review status: 0 of 3 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/kubectl/resource/visitor.go, line 644 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL4JIV0tp8--DmnnEQ:-KTL4JIWpJudq-2HnS9f:bf8oqu3) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/resource/visitor.go#L644)):*
> ```Go
> }
> 
> type FilterFunc func(info *Info, err error) (bool, error)
> ```

Comments for all this stuff to say what it does

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075)*

**Comment by mikedanese** [comment]
> General comment - this will result in the canonical result of apply to no longer live in source control since the label selectors define what will be deleted. Seems like this is very error prone - running apply on a subdirectory with prune will cause it to delete all sorts of things. At a minimum I think we need to print out what will be deleted and force the user to confirm.

@bgrant0607 @ghodss can you respond to @pwittrock 's above comment?

**Comment by mikedanese** [comment]
Review status: 0 of 1 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 95 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL-dV0KuBMMOvUyhxR:-KTLLvIXkrxrBjTKUAY8:btkeg46) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L95)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Ug.  Does the garbage collector handle this for us yet?
</details>
It needs to be implemented for all types which is not yet complete. cc @caesarxuchao 

---

*[pkg/kubectl/cmd/apply.go, line 98 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL-pE24pjujI6hatcx:-KTLLxbEHLxmmLonYXRa:bcy9imv) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L98)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> What does an empty value mean?  Delete everything by default?  We probably shouldn't permit using the default in that case
</details>
Empty value for a label selector means match everything here and everywhere else in kubectl.

---

*[pkg/kubectl/resource/builder.go, line 541 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KS5fBuUr91Y_YVDcOGa-r2-541:-KTLMsEMfGNC6fiY1iAg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a279a0483dcf556246ecff34a24db1567a7a2b9c/pkg/kubectl/resource/builder.go#L541)):*
<details><summary><i>Previously, mikedanese (Mike Danese) wrote…</i></summary>
> https://github.com/kubernetes/kubernetes/pull/32599
</details>
Done.

---

*[pkg/kubectl/resource/visitor.go, line 644 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL4JIV0tp8--DmnnEQ:-KTLMeLXMoPwbvNMdrXP:bu9ewdo) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/resource/visitor.go#L644)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Comments for all this stuff to say what it does
</details>
This was from a dependent pr. Reabased this commit out.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075)*

**Comment by caesarxuchao** [comment]
*[pkg/kubectl/cmd/apply.go, line 95 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTL-dV0KuBMMOvUyhxR:-KTLQRxpK0ZLgn_CfKTX:b-n794hf) ([raw file](https://github.com/kubernetes/kubernetes/blob/05d68d214503c025da4f03141b35cde5031a02e1/pkg/kubectl/cmd/apply.go#L95)):*
<details><summary><i>Previously, mikedanese (Mike Danese) wrote…</i></summary>
> It needs to be implemented for all types which is not yet complete. cc @caesarxuchao 
</details>
Yeah, GC is not enabled for all types yet.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075)*

**Comment by pwittrock** [comment]
cc @jml3on

**Comment by bgrant0607** [comment]
@mikedanese This should only prune resources directly created by apply, which it should be possible to deduce by looking for the configuration annotation.

**Comment by mikedanese** [comment]
@bgrant0607 ok, done.

**Comment by pwittrock** [comment]
FYI

>
pkg/kubectl/cmd/apply.go:317: undefined: annotations in annotations.LastAppliedConfigAnnotation

**Comment by pwittrock** [comment]
*[hack/make-rules/test-cmd.sh, line 1073 at r7](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTQkRPHrSo3mOxk6J-r:-KTQkRPJmDFrodCsLPXP:b-e9pdxe) ([raw file](https://github.com/kubernetes/kubernetes/blob/3ad8efc8e142526e4791ee75e26f1a3da88bf7ca/hack/make-rules/test-cmd.sh#L1073)):*
> ```Shell
>   # check wrong pod doesn't exist
>   output_message=$(! kubectl get pods a 2>&1 "${kube_flags[@]}")
>   kube::test::if_has_string "${output_message}" 'pods "a" not found'
> ```

what is the expected if you apply a deployment yaml in the same group that doesn't contain a pod?  Should it delete the pod?  e.g. usecase: changing the add-on type from a RC to a Deployment or something.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075)*

**Comment by pwittrock** [comment]
Review status: 0 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/kubectl/cmd/apply.go, line 166 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTQjukyp5U5A25WkNSp:-KTQjukzo0tBRaaFYCtk:b-r26xvb) ([raw file](https://github.com/kubernetes/kubernetes/blob/64bd62a8c1beed24ab0208dcd7dc1cbca1ad7620/pkg/kubectl/cmd/apply.go#L166)):*
> ```Go
> 		if info.Namespaced() {
> 			visitedNamespaces.Insert(info.Namespace)
> 			visitedNamespacedRESTMappings[info.Mapping.GroupVersionKind] = info.Mapping
> ```

Will this delete resource for which all resources of that mapping have been removed from the config files?

---

*[pkg/kubectl/cmd/apply.go, line 283 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-KTQgF9lPhSy1bxjmya3:-KTQgF9mNtUBNuiXhxNb:btq05nm) ([raw file](https://github.com/kubernetes/kubernetes/blob/64bd62a8c1beed24ab0208dcd7dc1cbca1ad7620/pkg/kubectl/cmd/apply.go#L283)):*
> ```Go
> }
> 
> type pruner struct {
> ```

Plz follow up later with comments that explain what all of this is about.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075)*

**Comment by mikedanese** [comment]
> Will this delete resource for which all resources of that mapping have been removed from the config files?

zihong mentioned this here: https://github.com/kubernetes/kubernetes/pull/33075#discussion_r80808989 . answer is no, we need to think about the api for that.

**Comment by pwittrock** [comment]
@k8s-bot unit test this

File: pkg/kubectl/cmd/apply.go [hunk 11]
```diff
@@ -55,7 +63,9 @@ var (
 		This resource will be created if it doesn't exist yet.
 		To use 'apply', always create the resource initially with either 'apply' or 'create --save-config'.
 
-		JSON and YAML formats are accepted.`)
+		JSON and YAML formats are accepted.
+		
+		Alpha Disclaimer: the --prune functionality is not yet complete. Do not use unless you are aware of what the current state is. See https://issues.k8s.io/34274.`)
```
Comment (mikedanese) [inline pkg/kubectl/cmd/apply.go]: @pwittrock

**Comment by MrHohn** [comment]
> This should only prune resources directly created by apply, which it should be possible to deduce by looking for the configuration annotation.

But is there a way to do so if user does want to prune those resources? 

Again take the Addon Manager as an use case. After upgrade the Addon Manager tries to use `apply --prune` for pruning the old resources, which is not created through `apply` nor `create --save-config`. It would be great if this command also supports this.

File: pkg/kubectl/cmd/apply.go (hunk 11 above)
Comment (pwittrock) [inline pkg/kubectl/cmd/apply.go]: LGTM

**Review by pwittrock — APPROVED** [review]

**Comment by pwittrock** [comment]
We should add a test that makes sure when --prune is not supplied, that pruning doesn't actually happen.  Also would be good to test that --all is actually required when no label selectors are used.

**Comment by pwittrock** [comment]
Reviewed 3 of 3 files at r6, 1 of 4 files at r9, 3 of 3 files at r10.
Review status: 1 of 4 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075)*

**Comment by pwittrock** [comment]
:lgtm:

---

Reviewed 1 of 3 files at r1, 1 of 2 files at r5, 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33075#-:-KTS0fNKe37ISFaUf1Vs:bnfp4nl)*

**Comment by mikedanese** [comment]
There are a few outstanding review comments that need to be fixed in followups.

**Comment by bgrant0607** [comment]
@MrHohn Deletion of resources that were never managed by kubectl apply is not the responsibility of kubectl apply. The addon manager will need to deal with that during the transition from the previous approach to the new approach.

**Comment by MrHohn** [comment]
@bgrant0607 Thanks for reply. Yes, in theory `kubectl apply` should not be responsible for this.

I'm now thinking about faking the `kubectl.kubernetes.io/last-applied-configuration` annotation and still utilize this feature to prune the old things.

**Comment by ghodss** [comment]
>> General comment - this will result in the canonical result of apply to no longer live in source control since the label selectors define what will be deleted. Seems like this is very error prone - running apply on a subdirectory with prune will cause it to delete all sorts of things. At a minimum I think we need to print out what will be deleted and force the user to confirm.

> @bgrant0607 @ghodss can you respond to @pwittrock 's above comment?

I don't understand this - apply --prune should *always* only delete objects with the last-applied-configuration annotation, right? So how would it delete things that it shouldn't?

**Comment by pwittrock** [comment]
Correct me if I am wrong on either of these:

Case 1:  Applying a sub-directory accidentally

IIUC this will delete everything from the root config dir except the things in this sub dir

```sh
cd /root/config/dir
kubectl apply -f ./
# Sometime later...
cd /root/config/dir/sub
kubectl apply -f .
```

Case 2: Applying 2 different directories in the same namespace

IIUC this will delete everything owned by team 1

```
# Person 1
kubectl apply -f /team/1/config/dir
# Person 2
kubectl apply -f /team/2/config/dir
```

Am I miss understanding the results of these actions?

**Comment by bgrant0607** [comment]
@pwittrock I assume that you meant those commands to specify `--prune`.

You are not wrong. However:

Your first case is just user error.

I'd expect different teams to use different namespaces. That was the original motivation for namespaces. There are many ways users can step on each other when multiple teams share the same namespace.

I'd recommend users use prune at the namespace level, without a label selector. However, I'd like all commands to support filtering via label selector, for additional control.

Additionally, this is not the only case of kubectl commands depending on command-line arguments. All of the generators do, as well: run, create secret, create configmap, etc. Users have asked for declarative (apply-compatible) ways of specifying such commands, despite the fact that it's easy to work around in most cases, such as using a script, make, nulecule, or some other tool. We'll need to decide whether to support that. I added a bullet in the roadmap about this a while ago. Perhaps the Kubefiles idea could be extended to support this if it looks compelling enough.

**Comment by pwittrock** [comment]
Yup I meant `--prune`.  Thanks.

**Comment by pwittrock** [comment]
The consequences of the first user error could be quite severe.  IMHO it is plausible that a user executes an apply command thinking they are in one directory when they are in fact in another directory.

I think it would be pretty simple to require a safeguard `.applyconfig` file to be present in the root directory and to only prune a directory if this file was found.  The config could also contain a specific annotation to apply to each resource it created.  This would prevent either case from accidentally occurring.  Applying a directory without this config would give an error.

Example workflow:

```sh
kubectl apply init /root/config directory --config-annotation "project=myproject"
# Do some work
kubectl apply -f /root/config directory
# All resources created with annotation "project=myproject"
# Delete some resources...
kubectl apply --prune -f /root/config directory
# All resources with the "project=myproject" annotation are deleted.  No messy label selectors need to be specified.
```
